We want to allow builds in chroots that lack /bin/sh.  Thus, system(3)
and popen(3) need to be tweaked to use the right shell.  For the bootstrap
glibc, we just use whatever `sh' can be found in $PATH.  The final glibc
instead uses the hard-coded absolute file name of `bash'.

The second part that touches fcntl-linux.h reverts a change from glibc-2.25
which would require the statically compiled glibc used in bootstrapping to
depend on the linux kernel headers.

--- a/sysdeps/posix/system.c
+++ b/sysdeps/posix/system.c
@@ -134,7 +134,7 @@ do_system (const char *line)
       INIT_LOCK ();
 
       /* Exec the shell.  */
-      (void) __execve (SHELL_PATH, (char *const *) new_argv, __environ);
+      (void) __execvpe (SHELL_NAME, (char *const *) new_argv, __environ);
       _exit (127);
     }
   else if (pid < (pid_t) 0)

--- b/libio/iopopen.c	2012-06-30 21:12:34.000000000 +0200
+++ b/libio/iopopen.c	2012-12-19 12:52:29.000000000 +0100
@@ -226,7 +226,7 @@ _IO_new_proc_open (fp, command, mode)
 	    _IO_close (fd);
 	}
 
-      _IO_execl ("/bin/sh", "sh", "-c", command, (char *) 0);
+      execlp ("sh", "sh", "-c", command, (char *) 0);
       _IO__exit (127);
     }
   _IO_close (child_end);

--- a/sysdeps/unix/sysv/linux/bits/fcntl-linux.h
+++ b/sysdeps/unix/sysv/linux/bits/fcntl-linux.h
@@ -318,7 +318,15 @@ struct f_owner_ex
 
  
 /* Flags for fallocate.  */
+# define FALLOC_FL_KEEP_SIZE           1 /* Don't extend size of file
+                                            even if offset + len is
+                                            greater than file size.  */
+# define FALLOC_FL_PUNCH_HOLE          2 /* Create a hole in the file.  */
+# define FALLOC_FL_COLLAPSE_RANGE      8 /* Remove a range of a file
+                                            without leaving a
+                                            hole.  */
+# define FALLOC_FL_ZERO_RANGE          16 /* Convert a range of a
+                                             file to zeros.  */
-# include <linux/falloc.h>
 
  
 /* File handle structure.  */
